internal/middleware/auth.go

```diff
@@ -57,3 +63,39 @@ func AuthenticateRebac(next http.Handler, jimm jujuapi.JIMM) http.Handler {
 		next.ServeHTTP(w, r.WithContext(ctx))
 	}), jimm)
 }
+
+// AuthenticateViaBasicAuth performs basic auth authentication and puts an identity in the request's context
+func AuthenticateViaBasicAuth(next http.Handler, jimm jujuapi.JIMM) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		// extract auth token
+		_, password, ok := r.BasicAuth()
+		if !ok {
+			w.WriteHeader(http.StatusUnauthorized)
+			_, _ = w.Write([]byte("authentication missing"))
+			return
+		}
+		user, err := jimm.LoginWithSessionToken(ctx, password)
+		if err != nil {
+			w.WriteHeader(http.StatusUnauthorized)
+			_, _ = w.Write([]byte("error authenticating the user"))
+			return
+		}
+		next.ServeHTTP(w, r.WithContext(WithUserContext(ctx, user)))
+	})
+}
+
+// GetUserFromContext extracts the user from the context.
+func GetUserFromContext(ctx context.Context) (*openfga.User, error) {
+	identity := ctx.Value(userContext{})
+	user, ok := identity.(*openfga.User)
+	if !ok {
+		return nil, errors.E("cannot extract user from context")
+	}
+	return user, nil
+}
+
+// WithUserContext sets the user into the context and return the context
+func WithUserContext(ctx context.Context, user *openfga.User) context.Context {
```

Comment:
suggestion: withIdentity

- user might not be a user, but also a service account
- we can omit the "context" bit in the name as we know the return type
- does not need to be exported as this is the only package that will use the unexported identityContextKey to put the identity in the context